src/mame/skeleton/hudson_poems.cpp

```diff
@@ -40,65 +52,666 @@ class hudson_poems_state : public driver_device
+	u32 m_tilemapunk[4];
+	u32 m_tilemapcfg[4];
+	u32 m_tilemapscr[4];
+	u32 m_tilemaphigh[4];
+
+	s32 m_hackcounter;
+
+	required_device<xtensa_device> m_maincpu;
+	required_device<palette_device> m_palette;
+	required_device<screen_device> m_screen;
+	required_device<gfxdecode_device> m_gfxdecode;
+	required_shared_ptr<u32> m_mainram;
+	required_shared_ptr<u32> m_palram;
 };
 
 void hudson_poems_state::machine_start()
 {
+	save_item(NAME(m_unktable));
+	save_item(NAME(m_unktableoffset));
+
+	save_item(NAME(m_spritegfxbase));
+	save_item(NAME(m_spritelistbase));
+
+	save_item(NAME(m_tilemapbase));
+	save_item(NAME(m_tilemapunk));
+	save_item(NAME(m_tilemapcfg));
+	save_item(NAME(m_tilemapscr));
+	save_item(NAME(m_tilemaphigh));
+
+	save_item(NAME(m_hackcounter));
 }
 
+
+
 void hudson_poems_state::machine_reset()
 {
-	m_maincpu->set_pc(0x20010058);
+	m_unktableoffset = 0;
+	m_hackcounter = 0;
+
+	for (int i = 0; i < 4; i++)
+	{
+		m_spritegfxbase[i] = m_tilemapbase[i] = m_tilemaphigh[i] = m_tilemapunk[i] = m_tilemapcfg[i] = m_tilemapscr[i] = 0;
+	}
 }
 
+/*
+You can unlock the test menu on boot by inputing a 3 step sequence while the Konami logo is being shown:
+Step 1 checks if you input white or yellow 2 times
+Step 2 checks if you input red or blue 2 times
+Step 3 checks if you input white or green 3 times
+
+These are not grouped inputs, so for step 1 you must input white x2 or yellow x2 instead of white x1 and yellow x1 for it to count.
+
+*/
+
 static INPUT_PORTS_START( hudson_poems )
+	PORT_START( "IN1" )
+	PORT_BIT( 0x00000002, IP_ACTIVE_LOW, IPT_BUTTON1 ) PORT_PLAYER(1) PORT_NAME("White")
+	PORT_BIT( 0x00000004, IP_ACTIVE_LOW, IPT_BUTTON2 ) PORT_PLAYER(1) PORT_NAME("Yellow (Select Up)")
+	PORT_BIT( 0x00000008, IP_ACTIVE_LOW, IPT_BUTTON3 ) PORT_PLAYER(1) PORT_NAME("Red (Ok)")
+	PORT_BIT( 0x00000010, IP_ACTIVE_LOW, IPT_BUTTON4 ) PORT_PLAYER(1) PORT_NAME("Blue (Select Down)")
+	PORT_BIT( 0x00000200, IP_ACTIVE_LOW, IPT_BUTTON5 ) PORT_PLAYER(1) PORT_NAME("Green")
 INPUT_PORTS_END
 
+void hudson_poems_state::draw_sprites(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect)
+{
+	int spritebase = (m_spritelistbase & 0x0003ffff) / 4;
+	gfx_element *gfx = m_gfxdecode->gfx(2);
+
+	for (int i = 0; i < 64; i++)
+	{
+		int tilebase;
+		u16 spriteword0 = m_mainram[(spritebase + i * 2) + 0] & 0xffff;
+		u16 spriteword1 = m_mainram[(spritebase + i * 2) + 0] >> 16;
+		u16 spriteword2 = m_mainram[(spritebase + i * 2) + 1] & 0xffff;
+		u16 spriteword3 = m_mainram[(spritebase + i * 2) + 1] >> 16;
+
+		const int x = (spriteword3 & 0x03ff);
+		const int y = (spriteword2 & 0x03ff);
+		int tilenum = spriteword1 & 0x03ff;
+		const int pal = (spriteword2 & 0x7c00)>>10;
+
+		// is it selecting from multiple tile pages (which can have different bases?) (probably from a register somewhere)
+		tilebase = (m_spritegfxbase[(spriteword0 & 0x0300)>>8] & 0x0003ffff) / 32; // m_spritegfxbase contains a full memory address pointer to RAM
```

Comment:
Can you move the declaration of `tilebase` down to where it’s initialised, and make `spriteword*` `const` since they’re just used for extracting meaningful fields?

Reply:
done